AVRO-2955: Avro-C: Support all Visual Studio versions from 2010 to 2019

kiyolee opened the pull request:
AVRO-2955: Avro-C: Support all Visual Studio versions from 2010 to 2019

Fix compiling using Visual Studio 2010|2013|2015|2017|2019.
Fix building using either cmake/ninja or cmake/visualstudio.
Fix testing using cmake/ninja.
Improve Windows compatibility.
All existing tests passed using all Visual Studio versions mentioned.

Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-2955
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
*** Imported a couple of files (getopt.h/getopt_long.c) from FreeBSD with modification for Windows. BSD licensed sources should be compatible with the license here.

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
*** No new functionality added. All existing tests passed.

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does
*** No new functionality added.

kiyolee committed:
AVRO-2955: Only VS2010 or before requires /TP. Use C11 for VS2019 or later.

kiyolee committed:
AVRO-2955: Enforcing that VS2010 or before is not supported.

kiyolee committed:
AVRO-2955: Conditional compile GCC only features.

kiyolee committed:
AVRO-2955: Fix missing typedef of ssize_t for MSVC.

kiyolee committed:
AVRO-2955: Newer MSVC has its own stdint.h and inttypes.h.

kiyolee committed:
AVRO-2955: Do not define snprintf as macro for VS2019 or later.

kiyolee committed:
AVRO-2955: Fix strrchr() returns const char* in newer MSVC.

kiyolee committed:
AVRO-2955: Fix MSVC does not support empty struct.

kiyolee committed:
AVRO-2955: Avoid non-Windows headers.

kiyolee committed:
AVRO-2955: Define _CRT_SECURE_NO_WARNINGS to reduce compiler warnings.

kiyolee committed:
AVRO-2955: MSVC offers _stricmp() instead of strcasecmp().

kiyolee committed:
AVRO-2955: Fix running tests with ninja.

kiyolee committed:
AVRO-2955: Use opendir()/readdir() from "msdirent.h" for Windows.

kiyolee committed:
AVRO-2955: Fix file mode.

kiyolee committed:
AVRO-2955: VS2013 or before does not support inline keyword.

kiyolee committed:
AVRO-2955: Fix C++ compliance for VS2010.

kiyolee committed:
AVRO-2955: Fix VS2010 support.

kiyolee committed:
AVRO-2955: Fix z_stream buffer pointer assignment.

martin-g commented:
@kiyolee The PR is messed up somehow. It contains many unrelated commits.
Also the following files have no licence headers:
- lang/c/src/getopt/getopt.h
- lang/c/src/getopt/getopt_long.c

kiyolee commented:
@martin-g Didn't aware the last pull rebase was messed up. Now fixed.
The following files are originated from FreeBSD and modified to work with MSVC:
- lang/c/src/getopt/getopt.h
- lang/c/src/getopt/getopt_long.c
The license per file does exist and are kept unchanged. What have I missed?

martin-g commented:
> The license per file does exist and are kept unchanged. What have I missed?

The RAT test fails because the new files do not have the ASLv2 license. 
Since you are copying files from elsewhere we need to add an entry to the NOTICE file about that.

kiyolee committed:
AVRO-2955: Build avromod and avropipe for Windows

Import getopt_long() from FreeBSD.

kiyolee committed:
AVRO-2955: Add copyright notice for getopt
